File: server/haskell/src/Record/Download.hs
module Record.Download (download) where

import Auth.JWT (authorizedUserId)
import Control.Monad.IO.Class (liftIO)
import Data.Int (Int64)
import qualified Data.Text as T
import qualified Data.Text.Lazy as TL
import Data.Time (UTCTime, defaultTimeLocale, formatTime)
import qualified Data.Vector as V
import qualified Hasql.Pool as HP
import qualified Hasql.Session as HS
import qualified Hasql.Statement as S
import qualified Hasql.TH as TH
import Network.HTTP.Types (status200, status404, status500)
import Record.Go (Color (Black, White))
import SGF (Game (..), toSGF)
import Web.Scotty
  ( ScottyM,
    get,
    param,
    raiseStatus,
    setHeader,
    status,
    text,
  )

type RecordRow = (Int64, T.Text, T.Text, T.Text, T.Text, Int64, Double, T.Text, T.Text, UTCTime)

selectRecord :: S.Statement (Int64, Int64) RecordRow
selectRecord =
  [TH.singletonStatement|
    select
    board_size :: int8, name :: text, black_player :: text, white_player :: text, comment :: text, handicap :: int8, komi :: float8, ruleset :: text, winner :: text, created :: timestamptz
    from record_record
    where
    owner_id = $1 :: int8 and id = $2 :: int8
  |]

type MoveRow = (Maybe Int64, T.Text)

selectMoves :: S.Statement Int64 (V.Vector MoveRow)
selectMoves =
  [TH.vectorStatement|
    select
    position :: int8?, color :: text
    from record_move
    where
    record_id = $1 :: int8
  |]

toColor :: T.Text -> Color
toColor color' = case color' of
  "B" -> Black
  "W" -> White
  _ -> White -- TODO????

fromColor :: Color -> String
fromColor color = case color of
  Black -> "B"
  White -> "W"

download :: HP.Pool -> ScottyM ()
download pool = get "/api/records/:recordId/download/" $ do
  userId <- authorizedUserId
  recordId <- param "recordId"
  recordSelect <- liftIO $ HP.use pool $ HS.statement (userId, recordId) selectRecord
  movesSelect <- liftIO $ HP.use pool $ HS.statement recordId selectMoves
  case (recordSelect, movesSelect) of
    (Right record, Right moves') -> do
      let moves = [(fmap fromIntegral pos', toColor color') | (pos', color') <- V.toList moves']
      let (size', name', blackPlayer, whitePlayer, comment', handicap, komi, _ruleset, winner', created) = record
      let size = fromIntegral size'
          name = if name' == "" then Nothing else Just $ T.unpack name'
          comment = if comment' == "" then Nothing else Just $ T.unpack comment'
          dateString = formatTime defaultTimeLocale "_%Y_%m_%d.sgf" created
          filename = case name of
            Just name'' -> name'' ++ dateString
            Nothing -> T.unpack blackPlayer ++ "_vs_" ++ T.unpack whitePlayer ++ dateString
          winner = case winner' of
            "B" -> Just Black
            "W" -> Just White
            _ -> Nothing
          game =
            Game
              { name,
                comment,
                handicap = fromIntegral handicap,
                komi,
                blackPlayer = T.unpack blackPlayer,
                whitePlayer = T.unpack whitePlayer,
                winner,
                size = size,
                moves
              }
      status status200
      setHeader "Content-Type" "application/x-go-sgf"
      setHeader "Content-Disposition" $ TL.pack $ "attachment; filename=\"" ++ filename ++ "\""
      text $ TL.pack $ toSGF game
    (Left (HP.SessionError (HS.QueryError _ _ (HS.ResultError (HS.UnexpectedAmountOfRows 0)))), _) -> do
      raiseStatus status404 "Does not exist."
    _ -> do
      raiseStatus status500 "DB error"

File: server/rust/src/record/download.rs
use axum::{
    extract::{Path, State},
    http::{
        header::{CONTENT_DISPOSITION, CONTENT_TYPE},
        HeaderMap, StatusCode,
    },
};
use chrono::{DateTime, Utc};
use std::sync::Arc;
use tokio_postgres::Client;

use crate::auth::UserId;
use crate::db::{query, query_one};
use crate::derow;
use crate::record::go::Pos;

pub async fn download(
    State(client): State<Arc<Client>>,
    UserId(user_id): UserId,
    Path(record_id): Path<i64>,
) -> Result<(StatusCode, HeaderMap, String), (StatusCode, &'static str)> {
    let row = query_one(&client, "SELECT id, board_size, name, black_player, white_player, comment, handicap, komi, ruleset, winner, created FROM record_record WHERE owner_id = $1 AND id = $2", &[&user_id, &record_id]).await?;
    let board_size: i32 = row.get("board_size");
    let created: DateTime<Utc> = row.get("created");
    let name: String = row.get("name");
    let black_player: String = row.get("black_player");
    let white_player: String = row.get("white_player");
    let comment: String = row.get("comment");
    let handicap: i32 = row.get("handicap");
    let komi: f64 = row.get("komi");
    let ruleset: String = row.get("ruleset");
    let winner: String = row.get("winner");

    let comment = if !comment.is_empty() {
        format!("GC[{}]", comment)
    } else {
        "".to_string()
    };
    let name = if !name.is_empty() {
        format!("GN[{}]", name)
    } else {
        "".to_string()
    };
    let winner = match winner.as_ref() {
        "U" => "Void",
        "B" => "B+R",
        "W" => "W+R",
        _ => {
            return Err((
                StatusCode::INTERNAL_SERVER_ERROR,
                "Error parsing the winner",
            ))
        }
    }
    .to_string();

    let rows = query(
        &client,
        "SELECT position, color FROM record_move WHERE record_id=$1 ORDER BY move ASC",
        &[&record_id],
    )
    .await?;
    let alphabet: Vec<&str> = vec![
        "a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r",
        "s", "t",
    ];
    let formatted_moves: String = rows
        .iter()
        .map(|row| derow!(row, {position: Option<i32>, color: &str}))
        .map(|(pos, color)| (pos.map(Pos::new), color))
        .map(|(pos, color)| (pos.map(|pos| pos.to_coord(board_size)), color))
        .map(|(pos, color)| {
            (
                match pos {
                    Some((x, y)) => format!(
                        "{}{}",
                        alphabet[TryInto::<usize>::try_into(x).unwrap()],
                        alphabet[TryInto::<usize>::try_into(y).unwrap()]
                    ),
                    None => "tt".to_string(),
                },
                color,
            )
        })
        .map(|(pos, color)| format!("{}[{}]", color, pos))
        .collect::<Vec<String>>()
        .join(";");
    let sgf_file = format!(
        "(;FF[4]CA[UTF-8]{}GM[1]{}HA[{}]KM[{}]PB[{}]PW[{}]RE[{}]SZ[{}];{})",
        comment,
        name,
        handicap,
        komi,
        black_player,
        white_player,
        winner,
        board_size,
        formatted_moves
    );
    let date_string = created.format("%Y_%m_%d");
    let filename = if !name.is_empty() {
        format!("{}_{}.sgf", name, date_string)
    } else {
        format!("{}_vs_{}_{}.sgf", black_player, white_player, date_string)
    };
    let mut headers = HeaderMap::new();
    headers.insert(CONTENT_TYPE, "application/x-go-sgf".parse().unwrap());
    headers.insert(
        CONTENT_DISPOSITION,
        format!("attachment; filename=\"{}\"", filename)
            .parse()
            .unwrap(),
    );
    Ok((StatusCode::OK, headers, sgf_file))
}
